### ISSUE TITLE
Implementation of Normalized Euclidean Distance (NED) in pytorch

### ISSUE THREAD

**Issue by brando90** [pytorch/pytorch#52005]
## 🚀 Feature
Implement an official version of normalized Euclidean distance. Reference of definition:
- https://stats.stackexchange.com/questions/136232/definition-of-normalized-euclidean-distance/498753?noredirect=1#comment937825_498753

## Motivation

A pytorch approved and numerically stable implementation of normalized euclidean distance

## Pitch

Current implementation:

```
def ned_torch(x1, x2, dim=1, eps=1e-8):
    """
    Normalized eucledian distance in pytorch.

    https://discuss.pytorch.org/t/how-does-one-compute-the-normalized-euclidean-distance-similarity-in-a-numerically-stable-way-in-a-vectorized-way-in-pytorch/110829
    https://stats.stackexchange.com/questions/136232/definition-of-normalized-euclidean-distance/498753?noredirect=1#comment937825_498753

    :param x1:
    :param x2:
    :param dim:
    :param eps:
    :return:
    """
    if x1.size(1) == 1:
        ned_2 = 0.5 * ((x1 - x2)**2 / (x1**2 + x2**2 + eps))
    else:
        # num_dims = len(x1.size())
        # dim = torch.tensor(range(1, num_dims))
        ned_2 = 0.5 * ((x1 - x2).var(dim=dim) / (x1.var(dim=dim) + x2.var(dim=dim) + eps))
    return ned_2 ** 0.5
```

test:

```
def test_ned():
    import torch.nn as nn

    # dim = 1  # apply cosine accross the second dimension/feature dimension

    k = 4  # number of examples
    d = 8  # dimension of feature space
    for d in range(1, d):
        x1 = torch.randn(k, d)
        x2 = x1 * 3
        print(f'x1 = {x1.size()}')
        ned_tensor = ned_torch(x1, x2)
        print(ned_tensor)
        print(ned_tensor.size())
```

## Alternatives

No alternative, NED is not in pytorch

## Additional context

Related:
- https://discuss.pytorch.org/t/how-does-one-compute-the-normalized-euclidean-distance-similarity-in-a-numerically-stable-way-in-a-vectorized-way-in-pytorch/110829

**Comment by brando90** [comment]
implementation fixed:

```
def ned_torch(x1: torch.Tensor, x2: torch.Tensor, dim=1, eps=1e-8) -> torch.Tensor:
    """
    Normalized eucledian distance in pytorch.

    Cases:
        1. For comparison of two vecs directly make sure vecs are of size [B] e.g. when using nes as a loss function.
            in this case each number is not considered a representation but a number and B is the entire vector to
            compare x1 and x2.
        2. For comparison of two batch of representation of size 1D (e.g. scores) make sure it's of shape [B, 1].
            In this case each number *is* the representation of the example. Thus a collection of reps
            [B, 1] is mapped to a rep of the same size [B, 1]. Note usually D does decrease since reps are not of size 1
            (see case 3)
        3. For the rest specify the dimension. Common use case [B, D] -> [B, 1] for comparing two set of
            activations of size D. In the case when D=1 then we have [B, 1] -> [B, 1]. If you meant x1, x2 [D, 1] to be
            two vectors of size D to be compare feed them with shape [D].

    https://discuss.pytorch.org/t/how-does-one-compute-the-normalized-euclidean-distance-similarity-in-a-numerically-stable-way-in-a-vectorized-way-in-pytorch/110829
    https://stats.stackexchange.com/questions/136232/definition-of-normalized-euclidean-distance/498753?noredirect=1#comment937825_498753
    """
    # to compute ned for two individual vectors e.g to compute a loss (NOT BATCHES/COLLECTIONS of vectorsc)
    if len(x1.size()) == 1:
        # [K] -> [1]
        ned_2 = 0.5 * ((x1 - x2).var() / (x1.var() + x2.var() + eps))
    # if the input is a (row) vector e.g. when comparing two batches of acts of D=1 like with scores right before sf
    elif x1.size() == torch.Size([x1.size(0), 1]):  # note this special case is needed since var over dim=1 is nan (1 value has no variance).
        # [B, 1] -> [B]
        ned_2 = 0.5 * ((x1 - x2)**2 / (x1**2 + x2**2 + eps)).squeeze()  # Squeeze important to be consistent with .var, otherwise tensors of different sizes come out without the user expecting it
    # common case is if input is a batch
    else:
        # e.g. [B, D] -> [B]
        ned_2 = 0.5 * ((x1 - x2).var(dim=dim) / (x1.var(dim=dim) + x2.var(dim=dim) + eps))
    return ned_2 ** 0.5

def nes_torch(x1, x2, dim=1, eps=1e-8):
    return 1 - ned_torch(x1, x2, dim, eps)
```

**Comment by mruberry** [comment]
One alternative would be providing an implementation like your example in a Github repo. 

I have two questions:

1) how would we make this more numerically stable than the implementation you've suggested?
2) is this function used in contexts that require that additional numerical stability (if there is any to be had)

**Comment by brando90** [comment]
> Thank you for this suggestion, @brando90. One alternative would be providing an implementation like your example in a Github repo.
> 
> I have two questions:
> 
> 1. how would we make this more numerically stable than the implementation you've suggested?
> 2. is this function used in contexts that require that additional numerical stability (if there is any to be had)

Hi!

1. The implementation I suggested work according to my code. I am unsure what else you require, is it just making a brando new repo with only the above code copy pasted?

2. I am no numerical stability expert so my hope is that someone that does know about this sees our code and checks it as fine if it is or suggests something if needs to be changed.


Thanks for the reply!

**Comment by mruberry** [comment]
> 1. The implementation I suggested works according to my tests. I am unsure what else is require, is it just making a brand new repo with only the above code and copy paste it?

That's one way to do it. Then people looking for an implementation could find one at your repo.

I think hosting a function like this externally is the best thing for now.

**Comment by mruberry** [comment]
Thanks @brando90! Let's see if we get additional requests or people start using the version in your repo.
